ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -1947,8 +1956,11 @@ private void updateStateForMessages(List<MessageInfo> messageEntries) throws Sto
     ListIterator<MessageInfo> messageEntriesIterator = messageEntries.listIterator();
     while (messageEntriesIterator.hasNext()) {
       MessageInfo messageInfo = messageEntriesIterator.next();
-      // for all the message info, we use most recent ref even for delete. since a deleted entry can be undeleted.
-      // ok to use most recent ref
+      // If the message is already deleted and DELETE is the final state of the blob, then we don't have
+      // to update the state for this message.
+      if (isDeleteFinalStateOfBlob && messageInfo.isDeleted()) {
+        continue;
```

Comment:
Should we add unit tests for this?

Reply:
We already have findEntriesSinceTest in the IndexTest file. I added a new parameter to that unit test class for this.